API: Add v2 Actions API

aokolnychyi opened the pull request:
This PR is the first step in reworking our actions. It introduces a set of interfaces in the API module that should be implemented by query engine integrations.

**Note**: new interfaces don't end with `...Action` to avoid conflicts with existing classes. I will gradually move action by action to the new API while keeping the compatibility (old actions will wrap new ones).

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * An API that should be implemented by query engine integrations for providing actions.
+ */
+public interface ActionsProvider {
```
I wanted to call it `Actions` but there is an existing class with that name in the same package :(

aokolnychyi replied:
I will deprecate the existing one once the implementation is ready.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * Instantiates an action to snapshot an existing table.
+   */
+  default SnapshotTable snapshotTable(String sourceTableIdent, String destTableIdent) {
```
One of the issues with the old API is that always assumed it operated on a single existing table. This method was offered as a static method there, which looked weird. In the new API, it is no longer a static method.

rdblue replied:
It's too bad we can't have a required method. I'd much rather not rely on order of arguments here and pass the destination name separately:

```
SparkActions.get().snapshotTable("a.b").as("a.c").execute();
```

aokolnychyi replied:
I think your concern is valid. Supporting `snapshotTable("db.tbl").as("db.tbl_2")` isn't ideal as we don't pass the required param right away but I'll be okay throwing an exception if the destination name is not set by calling `as`.
That will be a bit better than what we have now.

aokolnychyi replied:
Well, we could also default it to something like `db.tbl_iceberg_snapshot`.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * Instantiates an action to migrate an existing table to Iceberg.
+   */
+  default MigrateTable migrateTable(String tableIdent) {
```
I don't really like using Strings as identifiers as it requires implementations to parse it but I am not sure other options are better. I considered using `TableIdentifier` Iceberg defines but it is in the catalog module so it will be a bit weird to use it here. Also, parsing seems inevitable as there may be engine-specific parsing.

rdblue replied:
Yeah, I think a String is a reasonable choice here.

aokolnychyi replied:
Ok, let me resolve this one then.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ConfigurableAction.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @param <R> the Java type of the result produced by this action
+ */
+public interface ConfigurableAction<ThisT, R> extends Action<R> {
```
I am debating whether we need a separate interface. There are actions that extend both `ConfigurableAction` and `SnapshotUpdate` (e.g. `RewriteDataFiles`).

aokolnychyi replied:
Removed for now.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ConfigurableAction.java:
```diff
@@ -0,0 +1,46 @@
+   * @param value an option value
+   * @return this for method chaining
+   */
+  ThisT withOption(String name, String value);
```
There will be configs specific to engines. For example, whether to use `toLocalIterator` in Spark or not. It does not make sense to expose such a method in this API as other engines may not support it.

There are two approaches we can take:
- Support a map of options in the API.
- Have engine-specific interfaces that extend base action interfaces. For example, `ExpireSnapshotsSparkAction` will extend `ExpireSnapshotsAction`.

I went with the first option as it seems simpler. I can reconsider depending on what the community thinks.

RussellSpitzer replied:
I think this follows some of our discussion yesterday. I would prefer that any options that a user would not modify in 90% of use cases be an "option", I think engine specific things fall into that category. The example of `toLocalIterator` is a good one since it takes a much more advanced Spark user to understand the requirement so we should hide that in the public api exposed via any java IDE.

rymurr replied:
I don't love an opque map here but I see the purpose.

aokolnychyi replied:
Sounds like an intermediate consensus then. I'll keep it as is for now.

rdblue replied:
I think we could move this to the default Action interface if we wanted to make it a bit simpler.

aokolnychyi replied:
Do you mean `withOptions` or engine-specific methods, @rdblue?

aokolnychyi replied:
I'll resolve this one as we agreed in another thread.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ExpireSnapshots.java:
```diff
@@ -0,0 +1,114 @@
+  /**
+   * The action result that contains a summary of the execution.
+   */
+  interface Result {
```
I think representing results as interfaces will allow us to evolve them easier.

RussellSpitzer replied:
Is this a concern? I'm not sure we have had this issue yet. I don't think there is a real negative to doing this I just don't have a lot of imaginative thinking here.

aokolnychyi replied:
For example, `RemoveOrphanFiles` returns `List<String>` directly. It is impossible to change the return type without breaking the API. I think having a layer of abstraction would be safer. Who knows what will we come up with? What if we decide to report orphan metadata and data files separately?

pvary replied:
Your example should be a breaking change which is better highlighted by changing the API instead of hiding it behind an interface and depend on the documentation to describe the change.

But that is just an extreme example and I can come up with cases for both approaches, so no clean preference on my side.

aokolnychyi replied:
I meant the following interface:

```
interface Result {
  Iterable<String> orphanFileLocations();
}
```

Can safely evolve into (just an example, may not be the best idea):

```
interface Result {
  Iterable<String> orphanFileLocations(); // reports the same set which is a union of locations reported by methods below
  Iterable<String> orphanDataFileLocations();
  Iterable<String> orphanManifestLocations();
  Iterable<String> orphanManifestListLocations();
}
```

If we always returned just `List<String>`, we would be constrained to evolve the result type even when it is not a breaking change.

rdblue replied:
I prefer this approach because it puts the context in the API rather than in documentation. To get the locations, it is clear that they are the `orphanFileLocations` rather than a list of strings (if you don't have Javadoc to see what the strings are).

aokolnychyi replied:
Resolving this. Thanks to everybody who participated.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/RemoveOrphanFiles.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * Returns locations of orphan files.
+     */
+    Iterable<String> orphanFileLocations();
```
I am using `Iterable` vs some more specific implementations. Let me know what everybody thinks.

rymurr replied:
Is it possible that the orphaned files could be generated lazily? Would a `Stream` be more appropriate?

rdsr replied:
Lazy could make sense, [and some impls of Iterable could be lazy], but in this case, who owns responsibility releasing the underlying  resource?

RussellSpitzer replied:
Closeable iterable then?

aokolnychyi replied:
I think `CloseableIterable` may make sense. We use it mostly in `TableScan`. The only downside that it will complicate the usage a bit. Some of those iterables can be large so I'd want to avoid extra copies. Let me think this through.

aokolnychyi replied:
Well, I think it makes sense to use `ClosableIterable` in this particular case. At the same time, other places in this PR could continue to use `Iterable`.

Let me know what you think, @rymurr @rdsr @RussellSpitzer.

rdsr replied:
I was looking at the V1 Action impl for RemoveOrphanFiles. Seems like the orphan file list is computed upfornt. @aokolnychyi do you see the list being lazy in this particular case? My argument was mostly for lazy streams (or lazy impl of `Iterables`). If all our results are computed eagerly then returning a List is fine as that would remove the question of resource cleanup and ownership. I'm fine with `Iterables` also]

aokolnychyi replied:
We don't have a use case where this would be evaluated lazily yet. I can make it `ClosableIterable` if we think other implementations may benefit from this but it will definitely make using this API harder.

My motivation for changing the return type to `Iterable` is that it allows us to change the underlying representation. For example, we could migrate from `List` to `Set` in the internal implementation.

jackye1995 replied:
> Well, I think it makes sense to use `ClosableIterable` in this particular case. At the same time, other places in this PR could continue to use `Iterable` as it is unlikely they will benefit from lazy materialization.

Why? Cases like `Iterable<ManifestFile> rewrittenManifests()` can also benefit from using `CloseableIterable`. I feel if we want to do it for one, we should do it for all, because we cannot just assume one result might need to close resource, the other does not. If an engine does it for one action, it is likely that it will have this behavior for all actions.

With that being said, I agree that using `CloseableIterable` would make the interface harder to use. Why not just use `Iterable`, and let `Result` be the object to implement `close()`? If one engine has a closable iterable, it can be closed there.

aokolnychyi replied:
> Cases like Iterable<ManifestFile> rewrittenManifests() can also benefit from using CloseableIterable. I feel if we want to do it for one, we should do it for all, because we cannot just assume one result might need to close resource, the other does not. If an engine does it for one action, it is likely that it will have this behavior for all actions.

Given that it adds complexity, I feel like this should be considered on a case by case basis. For orphan files, it may make sense as we can stream the result but I am not sure about cases like `rewrittenManifests`. Most likely, we would need to materialize the list of manifests to rewrite at some point as we need to process all entries in them.

> Why not just use Iterable, and let Result be the object to implement close()?

I think this is a better idea than using `ClosebleIterable`. In this case, we probably should make all action results closable for consistency.

What does everybody on this thread think?

rymurr replied:
+1

aokolnychyi replied:
There is also an option of having engine-specific return types. However, that is going to complicate things further. It means we need engine-specific actions and to propagate `ThisT` everywhere in base actions. Plus, the return type may be substantially different from engine to engine.

rdblue replied:
Do we need this much complication? I don't think that we would necessarily want to make this lazily evaluated because we expect the set of orphan files to be small and a concrete type is easiest for a caller to handle. I have no problem with this being a `List`, but `Iterable` is also fine with me to be able to change it to a `Set` later.

aokolnychyi replied:
Well, it is hard for me to predict how useful that would be for other engines. We don't really use it in Spark and maybe if we treat these results as summaries, that should be fine. In the worst case, we can cast action to an engine specific type and call a method that returns a custom object. For example, we have `expire` that returns `DataFrame` in Spark.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ExpireSnapshots.java:
```diff
@@ -0,0 +1,114 @@
+ * Similar to {@link org.apache.iceberg.ExpireSnapshots} but may use a query engine to distribute
+ * parts of the work.
+ */
+public interface ExpireSnapshots extends ConfigurableAction<ExpireSnapshots, ExpireSnapshots.Result> {
```
As I wrote in the PR description, we don't have `Action` in the name for backward compatibility.

rdblue replied:
I think I prefer it like this anyway. Any time we can get rid of extra words, we should.

aokolnychyi replied:
Ok, then let's keep it this way as we don't really have alternatives.

aokolnychyi commented:
It would be great to get as many eyes as possible on this one. Especially, from folks working on query engines other than Spark.

@rdblue @RussellSpitzer @pvary @openinx @holdenk @shardulm94 @danielcweeks @rdsr

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * An action that rewrites data files.
+ */
+public interface RewriteDataFiles extends
```
This one will be reworked by @RussellSpitzer as he is working on a new proposal. So we may delay discussing this particular one.

RussellSpitzer replied:
Thanks, Sorry I haven't put this up yet but my goal is a more general action which allows many different types of underlying algorithms for compacting data files. Should have a pr up for discussion with some prototypes next week.

rymurr commented on api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * An API that should be implemented by query engine integrations for providing actions.
+ */
+public interface ActionsProvider {
```
One thing that I would be concerned about here is this interface grows forever with every new action we add. I can imagine some actions are going to be engine (or a subgroup of engines) specific as well. I wonder if its possible to have a more dynamic way of registering actions (eg an action catalog of sorts) or some categories and an interface per category?

aokolnychyi replied:
Could you elaborate a bit more on the idea, @rymurr? Would we load actions by name?

rymurr replied:
@aokolnychyi I think we could do either by name, enum or even by interface eg `actionCatalog.action(RewriteDataFiles.class)`. Presumably the catalog would need to expose a list of actions it supports as well. Not sure what others think on this idea? If configurations in configurable actions are engine configurations we may be able to add that to the catalog interface rather than each action?

aokolnychyi replied:
I am open to considering this idea but I am not sure it is going to help.

It sounds like our `SparkProcedures` class. Loading by name/enum/class, will most likely give us a parent action interface, not a specific interface to call. Also, we will probably need to change how arguments are passed. And each time we add a new action, we will have to add a new constant to the enum, which is more or less equivalent to adding a new method here with a default implementation.

I see this API kind of close to our `Table` API. That one also has a number of methods but it is pretty convenient. That said, I do get the concern of having a lot of methods in this API but we can address it differently. For example, we will support multiple compaction strategies. Instead of offering a separate action for each type, we can make the configuration flexible.

rdblue replied:
I agree with Anton here. If we were to load by name, we wouldn't get a specific interface to configure the action. We could get the interface if we loaded by passing in the interface, but then I don't see how we'd be much better off. The API would be awkward (you have to find and supply the action interface you want) and the benefit is avoiding a few methods here.

I think what Anton has in the PR is a good step forward instead.

rymurr replied:
agreed :-)

aokolnychyi replied:
Let me resolve this one to simplify the review. Thanks, @rymurr @rdblue!

rymurr reviewed (commented):
Overall I like it! Nice one @aokolnychyi

rdsr commented on api/src/main/java/org/apache/iceberg/actions/SnapshotUpdate.java:
```diff
@@ -0,0 +1,40 @@
+
+/**
+ * An action that produces snapshots. This interface contains common methods for all
+ * actions that create a new table {@link Snapshot}.
```
nit: new snapshot

rdblue replied:
I agree with this. It's more clear to say `new {@link Snapshot}` than `new table snapshot`.

aokolnychyi replied:
Done.

rdsr reviewed (commented):
+1 from me

jackye1995 commented on api/src/main/java/org/apache/iceberg/actions/SnapshotTable.java:
```diff
@@ -0,0 +1,64 @@
+   * @param value the value of the property to add
+   * @return this for method chaining
+   */
+  SnapshotTable withProperty(String key, String value);
```
nit: same comment as before, can be `addProperty`

aokolnychyi replied:
Replaced with `tableProperty`.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/MigrateTable.java:
```diff
@@ -0,0 +1,56 @@
+   * @param value the value of the property to add
+   * @return this for method chaining
+   */
+  MigrateTable withProperty(String key, String value);
```
I think we actually have the overwrite behavior here as if there is a property with the same name, it will be replaced.  Maybe, we should improve the doc?

```
Sets a table property in the newly created Iceberg table. Any properties with the same key will be overwritten.
```

Is that any better, @jackye1995?

jackye1995 replied:
Yeah that looks good to me!

aokolnychyi replied:
Done.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/ActionsProvider.java:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * Instantiates an action to remove orphan files.
+   */
+  default RemoveOrphanFiles removeOrphanFiles(Table table) {
```
If we are passing names that will be parsed to the other methods, why not load the table in the action for these as well?

aokolnychyi replied:
Will it require proper catalog config in the session?

rdblue commented on api/src/main/java/org/apache/iceberg/actions/ExpireSnapshots.java:
```diff
@@ -0,0 +1,114 @@
+   * Passes an alternative executor service that will be used for manifests and data files deletion.
+   * <p>
+   * If this method is not called, unnecessary manifests and data files will still be deleted using
+   * a single threaded executor service.
```
Single-threaded service or in the current thread?

aokolnychyi replied:
Done.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -0,0 +1,104 @@
+  RewriteDataFiles filter(Expression expr);
+
+  /**
+   * Enables or disables case sensitive expression binding.
```
Should this mention the default?

aokolnychyi replied:
Fixed.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -0,0 +1,104 @@
+  RewriteDataFiles outputSpecId(int specId);
+
+  /**
+   * Specify the target rewrite data file size in bytes
```
Mention that the default is the table's target size.

aokolnychyi replied:
Done.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -0,0 +1,104 @@
+     * Returns rewritten data files.
+     */
+    Iterable<DataFile> rewrittenDataFiles();
+    /**
```
Nit: missing newline between methods.

aokolnychyi replied:
Fixed.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java:
```diff
@@ -0,0 +1,69 @@
+    SnapshotUpdate<RewriteManifests, RewriteManifests.Result> {
+
+  /**
+   * Rewrites manifests for a given spec id.
```
I think this should mention the default value, which is the table's default spec ID.

aokolnychyi replied:
Fixed.

rdblue commented on api/src/main/java/org/apache/iceberg/actions/SnapshotUpdate.java:
```diff
@@ -0,0 +1,40 @@
+   * @param value a String property value
+   * @return this for method chaining
+   */
+  ThisT set(String property, String value);
```
In other places, we use `with` and descriptive name (option, property). Should this be `withSnapshotProperty`?

aokolnychyi replied:
Good point, this has to be changed.

I am actually no longer sure having `with` prefix is a good idea as we have it for some configs but not others. Shall we consider dropping `with` from names? Then we will have `snapshotProperty`, `tableProperties`, `options`.

```
actions.rewriteManifests()
    .stagingLocation("custom-staging-location")
    .option("use-caching", "true")
    .snapshotProperty("label", "value")
    .execute();
```

```
actions.removeOrphanFiles()
    .location("custom-location")
    .option("use-caching", "true")
    .execute();
```

```
actions.snapshot("db.tbl")
    .as("db.tbl_2")
    .option("use-caching", "true")
    .location("custom-location")
    .tableProperties(props)
    .execute();
```

```
actions.migrate("db.tbl")
    .option("use-caching", "true")
    .tableProperty("p1", "v1")
    .tableProperty("p2", "v2")
    .execute();
```

aokolnychyi replied:
cc @RussellSpitzer who developed the original snapshot/migrate API.

aokolnychyi replied:
I've switched to the new approach.

RussellSpitzer replied:
I have no strong opnions about this although I do think we use the preposition-less(?) names with accessors usually rather than setters.

rdblue commented:
Overall this looks like a good step. Thanks @aokolnychyi! I think we just need to fix a few minor things and then it's ready to go.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/Action.java:
```diff
@@ -0,0 +1,62 @@
+ * @param <ThisT> the child Java API class, returned by method chaining
+ * @param <R> the Java type of the result produced by this action
+ */
+public interface Action<ThisT, R> {
```
This interface is still implemented by old actions that we will deprecate.

rdblue replied:
We intend to remove this eventually?

aokolnychyi replied:
I am not sure. I think it is reasonable to have a common parent even for new actions. What are your thoughts?

rdblue replied:
Yeah, I think it's reasonable to have a common parent. Your comment made me think you planned to remove it.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/ExpireSnapshots.java:
```diff
@@ -0,0 +1,114 @@
+ * Similar to {@link org.apache.iceberg.ExpireSnapshots} but may use a query engine to distribute
+ * parts of the work.
+ */
+public interface ExpireSnapshots extends Action<ExpireSnapshots, ExpireSnapshots.Result> {
```
I don't have `<ThisT>` in `ExpireSnapshots` as we don't expect engine-specific interfaces.

aokolnychyi commented:
I've addressed some of the comments. Still open questions:

- Having `options` in the `Action` directly. That seems simpler than having `ConfigurableAction`.
- Accepting Strings instead of Tables in all actions so that each action loads the table. My worry is that it will require a proper catalog config for resolution. I think most people will leverage SQL extensions for simple cases so it feels more flexible to pass the `Table` object. However, I agree it is not consistent with `snapshotTable` and `migrateTable`.
- Shall we make action results closeable? Seems like that would really complicate the usage so maybe we can skip doing that unless needed by other query engines.
- Using `with` prefix for some method names. I'd say let's not use it to be consistent.

rdblue commented:
> Having options in the Action directly. That seems simpler than having ConfigurableAction.

+1

> Accepting Strings instead of Tables in all actions so that each action loads the table.

I actually like that some accept a `Table`. I'm trying to think of a better way to structure the ones that take strings. I agree with the idea that it is easier to use that way, but the problem is that we need a way to parse the string into an identifier and to load a catalog. Maybe we should have a version that accepts a catalog and an identifier? I'm not sure what to do here.

> Shall we make action results closeable?

Let's skip it for now if we don't need it.

> Using with prefix for some method names. I'd say let's not use it to be consistent.

+1 for fewer words.

aokolnychyi commented:
>  I'm trying to think of a better way to structure the ones that take strings. I agree with the idea that it is easier to use that way, but the problem is that we need a way to parse the string into an identifier and to load a catalog...

That's the part I am not sure about. I considered using Iceberg's table identifier but is in the `catalog` package, which makes it weird. Plus, one identifier will point to a non-Iceberg table. Apart from that, only implementations will parse identifiers and the syntax could? actually vary from engine to engine. It is unlikely we can make catalog loading and identifier parsing generic.

aokolnychyi commented on api/src/main/java/org/apache/iceberg/actions/RewriteDataFiles.java:
```diff
@@ -0,0 +1,110 @@
+   * @param splitLookback number of "bins" considered when trying to pack the next file split into a task.
+   * @return this for method chaining
+   */
+  RewriteDataFiles splitLookback(int splitLookback);
```
I think we may want to hide this in options during the redesign, @RussellSpitzer.

aokolnychyi commented:
I think the only open point on this PR is whether we have any better ideas around snapshot/migrate methods. I think implementations like `SparkActions` can offer methods that accept Spark `Catalog` and `Identifier`. That's what we are going to use in our procedures. We may consider either exposing such methods alongside the methods that accept strings or we can introduce type parameters in `ActionProvider`.

```
public interface ActionsProvider<C, I> {

  default SnapshotTable snapshotTable(C sourceCatalog, I sourceIdent) {
    throw new UnsupportedOperationException(this.getClass().getName() + " does not implement snapshotTable");
  }

  default MigrateTable migrateTable(C catalog, I ident) {
    throw new UnsupportedOperationException(this.getClass().getName() + " does not implement migrateTable");
  }
```

aokolnychyi commented:
Could everyone please let me know their thoughts? I'd like to finish with this PR and submit follow-ups.

aokolnychyi commented:
Actually, we may not need the catalog. We can simply use the identifier.

aokolnychyi commented:
Well, maybe using just identifiers won't be sufficient as the catalog resolution can depend on the context. In case of procedures, we default the catalog to the procedure catalog, not the default catalog.

aokolnychyi commented:
Here is an example of what I mean:

```
public interface ActionsProvider<I> {

  default SnapshotTable<I> snapshotTable(I sourceTableIdent) {
    ...
  }

  default MigrateTable<I> migrateTable(I tableIdent) {
    ...
  }

 default RewriteManifests rewriteManifests(I tableIdent) {
    ...
  }

  default RewriteManifests rewriteManifests(Table table) {
    ...
  }

  ...
}

class SparkActions implements ActionsProvider<org.apache.spark.sql.connector.Identifier> {
}
```

RussellSpitzer commented:
@aokolnychyi Does that mean the caller needs to append the catalog to the identifier based on it's call context?

Ie if i am running Migrate Table Procedure from my_catalog and I get database.table as the arg, then the procedure makes a new identifier my_catalog.database.table and passes it through?

aokolnychyi commented:
> Ie if i am running Migrate Table Procedure from my_catalog and I get database.table as the arg, then the procedure makes a new identifier my_catalog.database.table and passes it through?

Well, that would work but not very clean. We can keep Catalog and Identifier separate as well. The main idea is to add params to `ActionsProvider` and use query engine specific classes instead of strings.

RussellSpitzer commented:
It seems rather complicated either way, although with a single Type parameter we could still pass a parameterized Pair (Catalog + Identifier)

aokolnychyi commented:
The problem with using Spark 3 `Identifier` is that Spark does not actually expose it anywhere except the catalog implementations, meaning we don't really have a good engine API to expose for our users.

Using our `TableIdentifier` to refer to non-Iceberg tables does not seem ideal either but let me think about it.

aokolnychyi commented:
We should probably pick between strings and Iceberg identifiers. I lean towards strings as we sometimes refer to non-Iceberg tables but it is not a strong opinion.

rdblue commented:
Let's use Strings for now. That's what we already do in the actions and we didn't have a problem with it before.

aokolnychyi commented:
Let me know if there are any other comments I should address.

rdblue commented:
Looks good to me!

aokolnychyi commented:
Thanks everyone for reviewing, @RussellSpitzer @rdsr @rymurr @jackye1995 @pvary @rdblue!
